lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1534,10 +1555,11 @@ bool TypeChecker::coerceParameterListToType(ParameterList *P, ClosureExpr *CE,
                  diag::param_type_non_materializable_tuple, ty);
       }
     }
-    
-    if (param->isInvalid())
-      param->overwriteType(ErrorType::get(Context));
-    else
+
+    // If contextual type is invalid and we have a valid argument type
+    // trying to coerce argument to contextual type would mean erasing
+    // valuable diagnostic information.
+    if (isValidType(ty) || shouldOverwriteParam(param))
       param->overwriteType(ty);
```

Comment:
I think it's safe actually to leave the hasError result as it is because the behavior related to `overwriteType` is the same in all of the param type error cases iff: **a.** we have explicit type set and contextual is not valid (but instead of only checking for errors it checks for all problems e.g. errors/unresolved/type vars) and **b.** if there is a problem later on with parameter type we always `overwrite` it, regardless of contextual type being valid or not, to replace error/typevar types with unresolved types. I agree that logic here is kind of confusing, but I think with these changes it's a bit less so...